sorting-algorithms.js

```diff
@@ -1,102 +1,129 @@
   while (leftIndex <= rightIndex) {
     if (array[leftIndex] > array[rightIndex]) {
-      swapAndTrackSteps(array, leftIndex, rightIndex, steps)
+      swapAndTrackSteps(array, leftIndex, rightIndex, steps);
     }
     if (array[leftIndex] <= array[pivotIndex]) {
-      leftIndex++
+      leftIndex++;
     }
     if (array[rightIndex] >= array[pivotIndex]) {
-      rightIndex--
+      rightIndex--;
     }
   }
-  swapAndTrackSteps(array, pivotIndex, rightIndex, steps)
+  swapAndTrackSteps(array, pivotIndex, rightIndex, steps);
   const isLeftSubarraySmaller =
-    rightIndex - 1 - startIndex < endIndex - leftIndex - 1
+    rightIndex - 1 - startIndex < endIndex - leftIndex - 1;
   if (isLeftSubarraySmaller) {
-    quickSortHelper(array, startIndex, rightIndex - 1, steps)
-    quickSortHelper(array, rightIndex + 1, endIndex, steps)
+    quickSortHelper(array, startIndex, rightIndex - 1, steps);
+    quickSortHelper(array, rightIndex + 1, endIndex, steps);
   } else {
-    quickSortHelper(array, rightIndex + 1, endIndex, steps)
-    quickSortHelper(array, startIndex, rightIndex - 1, steps)
+    quickSortHelper(array, rightIndex + 1, endIndex, steps);
+    quickSortHelper(array, startIndex, rightIndex - 1, steps);
   }
 }
 
 export function selectionSort(array) {
-  let steps = []
+  let steps = [];
   for (var i = 0; i < array.length - 1; i++) {
-    let minIndex = i
+    let minIndex = i;
     for (var j = i; j < array.length; j++) {
       if (array[minIndex] > array[j]) {
-        minIndex = j
+        minIndex = j;
       }
     }
-    swapAndTrackSteps(array, i, minIndex, steps)
+    swapAndTrackSteps(array, i, minIndex, steps);
   }
   // Completed sort with no swap positions
   // to draw the final step
-  steps.push({ array: [...array] })
-  return [array, steps]
+  steps.push({ array: [...array] });
+  return [array, steps];
 }
 
 function swap(array, i, j) {
-  const temp = array[i]
-  array[i] = array[j]
-  array[j] = temp
+  const temp = array[i];
+  array[i] = array[j];
+  array[j] = temp;
 }
 
 function swapAndTrackSteps(array, i, j, steps) {
-  swap(array, i, j)
+  swap(array, i, j);
   steps.push({
     array: [...array],
     swapPositions: [i, j]
-  })
+  });
+}
+//Merge Sort
+export function mergesort(arr) {
+  if (arr.length < 2) {
+    return arr;
+  } else {
+    var midpoint = parseInt(arr.length / 2);
+    var leftArr = arr.slice(0, midpoint);
+    var rightArr = arr.slice(midpoint, arr.length);
+    return merge(mergesort(leftArr), mergesort(rightArr));
+  }
+}
+
+function merge(leftArr, rightArr) {
+  var sortedArr = [];
+  while (leftArr.length && rightArr.length) {
+    if (leftArr[0] <= rightArr[0]) {
+      sortedArr.push(leftArr[0]);
+      leftArr = leftArr.slice(1);
+    } else {
+      sortedArr.push(rightArr[0]);
+      rightArr = rightArr.slice(1);
+    }
+  }
+  while (leftArr.length) sortedArr.push(leftArr.shift());
+  while (rightArr.length) sortedArr.push(rightArr.shift());
+  return sortedArr;
```

Comment:
This implementation of merge sort looks good, but it doesn't keep track of the steps to be able to draw in the canvas, which is the problem we need to solve here.

So basically, if you see other implementations of the algorithms in the file, I keep track of the `step`s [here](https://github.com/nicodelpiano/sorting-algorithms-visualizer/blob/003c1d327a3a4158ce21fd75cd269d7fc6a8bcc9/sorting-algorithms.js#L74) and then we have to figure out how to use `swapAndTrackSteps` function to properly save the steps.

Reply:
Okay will look into this and resolve it

Reply:
Hello 
Actually I am not able to properly figure out the use of this swap function and hoe to integrate it with current merge sort code so, can you please help me with it